compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUSplitReduction.cpp

```diff
@@ -24,6 +24,129 @@ namespace mlir {
 namespace iree_compiler {
 namespace {
 
+/// Make sure that
+/// - the pass has not been applied before
+/// - has tensor semantics
+/// - number of reduction loops == 1
+/// - has exactly 1 output
+/// - index map has only projected permutations
+/// - is a linalg generic op
+/// - has exactly 1 input
+/// - if enableReductionReordering is not set, then operand is an int
+/// - innermost dimension of the input operand is reduction
+/// TODO: support named ops, numInputs > 1, and modify lastDim check below
+/// accordingly. If fpReductionReordering is not enabled by default, it must
+/// be an integer or index type to proceed to allow associative reordering.
+LogicalResult splitReductionPrecondition(Operation *op,
+                                         bool fpReductionReordering) {
+  linalg::LinalgOp linalgOp = cast<linalg::LinalgOp>(op);
+
+  if (!linalgOp.hasTensorSemantics()) {
+    LLVM_DEBUG(llvm::dbgs() << "doesn't have tensor semantics\n");
+    return failure();
+  }
+  if (linalgOp.getNumReductionLoops() != 1) {
+    LLVM_DEBUG(llvm::dbgs() << "number of reduction loops != 1\n");
+    return failure();
+  }
+  if (linalgOp.getNumDpsInits() != 1) {
+    LLVM_DEBUG(llvm::dbgs() << "doesn't have exactly 1 output\n");
+    return failure();
+  }
+  if (!linalgOp.hasOnlyProjectedPermutations()) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "index map doesn't have only projected permutations\n");
+    return failure();
+  }
+  if (!isa<linalg::GenericOp>(op)) {
+    LLVM_DEBUG(llvm::dbgs() << "is not a generic op\n");
+    return failure();
+  }
+  if (linalgOp.getNumDpsInputs() != 1) {
+    LLVM_DEBUG(llvm::dbgs() << "doesn't have exactly 1 input\n");
+    return failure();
+  }
+
+  auto elemType =
+      getElementTypeOrSelf(linalgOp.getDpsInputOperand(0)->get().getType());
+  if (!(fpReductionReordering || elemType.isIntOrIndex())) {
+    LLVM_DEBUG(
+        llvm::dbgs()
+        << "skipped because reduction reordering on FP is not enabled.\n");
+    return failure();
+  }
+  return success();
+}
+
+/// Converts an inner-reduction into outer reduction + inner-parallel dimension,
+/// followed by simple inner reduction.
+LogicalResult splitReductionImpl(Operation *op, int64_t size,
+                                 RewriterBase &rewriter) {
+  IRRewriter::InsertionGuard g(rewriter);
+  rewriter.setInsertionPointAfter(op);
+  linalg::LinalgOp linalgOp = cast<linalg::LinalgOp>(op);
+
+  SmallVector<unsigned> dims;
+  linalgOp.getReductionDims(dims);
+  AffineMap map =
+      linalgOp.getMatchingIndexingMap(linalgOp.getDpsInputOperand(0));
+  unsigned lastIdx = map.getNumResults() - 1;
+  unsigned lastDim = map.getDimPosition(lastIdx);
+  if (lastDim != dims[0]) {
+    LLVM_DEBUG(
```

Comment:
I think you can replace LLVM_DEBUG this with `rewriter.notifyMatchError`

Reply:
not for this place.. it should get skipped if it does not meet this condition. let me refactor a bit for the case